File: app/server/analytics/digest.ts
import { Correlation, Meal, Symptom, UserSettings } from "@shared/schema";
import { addDays, localDate } from "./time";
import { clusterFlares, intensityOf, onsetHours } from "./flares";
import { COVERAGE_SLOTS } from "./coverage";
import { catalogItemByKey, catalogItemByName } from "@shared/symptomCatalog";

export type DayTrend = {
  date: string;
  weekday: string;
  index: number;
  occurrences: number;
  maxIntensity: number;
  level: "zero" | "moderate" | "high";
};

export type SymptomCard = {
  name: string;
  catalogKey: string | null;
  emoji: string | null;
  occurrences: number;
  shareOfWeek: number;
  avgSeverity10: number;
  avgDurationMinutes: number | null;
  peakDay: string | null;
  vsBaseline: "up" | "same" | "down";
  topTriggers: string[];
};

export type WeeklyDigest = {
  weekStart: string;
  weekEnd: string;
  previousWeekStart: string;
  hasNextWeek: boolean;
  trends: {
    index: number;
    baselineIndex: number;
    deltaVsBaselinePercent: number;
    previousWeekIndex: number;
    changeVsPreviousPercent: number;
    occurrences: number;
    flares: number;
    discomfortFreeDays: number;
    severeDays: number;
    severePeakDay: string | null;
    mealLogDepth: number;
    dataCompleteness: "complete" | "partial" | "empty";
    days: DayTrend[];
  };
  symptoms: {
    total: number;
    baselinePerWeek: number;
    vsBaseline: "up" | "same" | "down";
    distinct: number;
    distribution: { name: string; catalogKey: string | null; count: number; share: number }[];
    cards: SymptomCard[];
    onsetWindows: { under1h: number; from1to3h: number; over3h: number; unmatched: number };
  };
};

const WEEKDAYS = ["Sun", "Mon", "Tue", "Wed", "Thu", "Fri", "Sat"];
function weekdayOf(date: string): string {
  const [y, m, d] = date.split("-").map(Number);
  return WEEKDAYS[new Date(Date.UTC(y, m - 1, d))!.getUTCDay()];
}

export function weekDates(weekStart: string): string[] {
  return Array.from({ length: 7 }, (_, i) => addDays(weekStart, i));
}

/** Discomfort index for one day: max intensity × 2 (0 when nothing logged). */
function dayIndex(symptoms: Symptom[]): number {
  return symptoms.length === 0 ? 0 : Math.max(...symptoms.map(intensityOf)) * 2;
}

export function levelFor(index: number): DayTrend["level"] {
  if (index === 0) return "zero";
  return index >= 5 ? "high" : "moderate";
}

function round1(n: number): number {
  return Math.round(n * 10) / 10;
}

function compare(value: number, baseline: number): "up" | "same" | "down" {
  if (baseline === 0) return value === 0 ? "same" : "up";
  const ratio = value / baseline;
  if (ratio > 1.2) return "up";
  if (ratio < 0.8) return "down";
  return "same";
}

/**
 * Weekly digest for the 7 days starting at `weekStart`. `symptoms` and
 * `meals` are the user's full history so baselines can be computed.
 */
export function computeWeeklyDigest(
  meals: Meal[],
  symptoms: Symptom[],
  correlations: Correlation[],
  weekStart: string,
  tz: string,
  settings: UserSettings,
  today: string,
): WeeklyDigest {
  const dates = weekDates(weekStart);
  const weekEnd = dates[6];
  const inWeek = new Set(dates);

  const symptomsByDay = new Map<string, Symptom[]>();
  for (const s of symptoms) {
    const d = localDate(s.timestamp, tz);
    symptomsByDay.set(d, [...(symptomsByDay.get(d) ?? []), s]);
  }
  const mealsByDay = new Map<string, Meal[]>();
  for (const m of meals) {
    const d = localDate(m.timestamp, tz);
    mealsByDay.set(d, [...(mealsByDay.get(d) ?? []), m]);
  }

  // Trends
  const days: DayTrend[] = dates.map((date) => {
    const list = symptomsByDay.get(date) ?? [];
    const index = dayIndex(list);
    return { date, weekday: weekdayOf(date), index, occurrences: list.length, maxIntensity: list.length ? Math.max(...list.map(intensityOf)) : 0, level: levelFor(index) };
  });
  const weekIndex = days.reduce((sum, d) => sum + d.index, 0) / 7;

  // Baseline: mean daily index over every day since the first log
  const allDates = [...new Set([...symptomsByDay.keys(), ...mealsByDay.keys()])].sort();
  let baselineIndex = 0;
  if (allDates.length > 0) {
    const first = allDates[0];
    const last = today > weekEnd ? today : weekEnd;
    let cursor = first;
    let total = 0;
    let count = 0;
    while (cursor <= last && count < 3660) {
      total += dayIndex(symptomsByDay.get(cursor) ?? []);
      count++;
      cursor = addDays(cursor, 1);
    }
    baselineIndex = count ? total / count : 0;
  }

  const previousWeekStart = addDays(weekStart, -7);
  const previousIndex = weekDates(previousWeekStart).reduce((sum, d) => sum + dayIndex(symptomsByDay.get(d) ?? []), 0) / 7;

  const weekSymptoms = dates.flatMap((d) => symptomsByDay.get(d) ?? []);
  const weekMeals = dates.flatMap((d) => mealsByDay.get(d) ?? []);
  const flares = clusterFlares(weekSymptoms);
  const severe = days.filter((d) => d.level === "high");
  const peak = [...days].sort((a, b) => b.index - a.index)[0];

  const slotsLogged = dates.reduce((sum, d) => {
    const types = new Set((mealsByDay.get(d) ?? []).map((m) => m.mealType));
    return sum + COVERAGE_SLOTS.filter((slot) => types.has(slot)).length;
  }, 0);

  const loggedDays = dates.filter((d) => (mealsByDay.get(d)?.length ?? 0) > 0 || (symptomsByDay.get(d)?.length ?? 0) > 0).length;
  const elapsedDays = dates.filter((d) => d <= today).length;
  const dataCompleteness = loggedDays === 0 ? "empty" : loggedDays >= elapsedDays ? "complete" : "partial";

  // Symptoms section
  const byName = new Map<string, Symptom[]>();
  for (const s of weekSymptoms) byName.set(s.name, [...(byName.get(s.name) ?? []), s]);

  const historyWeeks = Math.max(1, allDates.length ? Math.ceil((daysBetween(allDates[0], today) + 1) / 7) : 1);
  const baselinePerWeek = symptoms.length / historyWeeks;
  const historyByName = new Map<string, number>();
  for (const s of symptoms) historyByName.set(s.name, (historyByName.get(s.name) ?? 0) + 1);

  const cards: SymptomCard[] = [...byName.entries()]
    .map(([name, list]) => {
      const catalog = catalogItemByKey(list[0].catalogKey) ?? catalogItemByName(name);
      const durations = list.map((s) => s.durationMinutes).filter((d): d is number => d != null);
      const perDay = new Map<string, number>();
      for (const s of list) {
        const d = localDate(s.timestamp, tz);
        perDay.set(d, (perDay.get(d) ?? 0) + 1);
      }
      const peakDay = [...perDay.entries()].sort((a, b) => b[1] - a[1] || a[0].localeCompare(b[0]))[0]?.[0] ?? null;
      const triggers = correlations
        .filter((c) => c.symptomName === name && c.dimension === "ingredient")
        .sort((a, b) => b.confidence - a.confidence)
        .slice(0, 3)
        .map((c) => c.foodName);
      return {
        name,
        catalogKey: list[0].catalogKey ?? catalog?.key ?? null,
        emoji: catalog?.emoji ?? null,
        occurrences: list.length,
        shareOfWeek: round1(list.length / weekSymptoms.length * 100) / 100,
        avgSeverity10: round1(list.reduce((sum, s) => sum + intensityOf(s) * 2, 0) / list.length),
        avgDurationMinutes: durations.length ? Math.round(durations.reduce((a, b) => a + b, 0) / durations.length) : null,
        peakDay,
        vsBaseline: compare(list.length, (historyByName.get(name) ?? 0) / historyWeeks),
        topTriggers: triggers,
      };
    })
    .sort((a, b) => b.occurrences - a.occurrences || a.name.localeCompare(b.name));

  const onsetWindows = { under1h: 0, from1to3h: 0, over3h: 0, unmatched: 0 };
  for (const s of weekSymptoms) {
    const h = onsetHours(meals, s.timestamp, settings.correlationWindowHours);
    if (h == null) onsetWindows.unmatched++;
    else if (h < 1) onsetWindows.under1h++;
    else if (h <= 3) onsetWindows.from1to3h++;
    else onsetWindows.over3h++;
  }

  return {
    weekStart,
    weekEnd,
    previousWeekStart,
    hasNextWeek: addDays(weekStart, 7) <= today,
    trends: {
      index: round1(weekIndex),
      baselineIndex: round1(baselineIndex),
      deltaVsBaselinePercent: baselineIndex > 0 ? Math.round((weekIndex - baselineIndex) / baselineIndex * 100) : 0,
      previousWeekIndex: round1(previousIndex),
      changeVsPreviousPercent: previousIndex > 0 ? Math.round((weekIndex - previousIndex) / previousIndex * 100) : 0,
      occurrences: weekSymptoms.length,
      flares: flares.length,
      discomfortFreeDays: days.filter((d) => d.index === 0).length,
      severeDays: severe.length,
      severePeakDay: peak && peak.index > 0 ? peak.date : null,
      mealLogDepth: Math.round(slotsLogged / 21 * 100) / 100,
      dataCompleteness,
      days,
    },
    symptoms: {
      total: weekSymptoms.length,
      baselinePerWeek: round1(baselinePerWeek),
      vsBaseline: compare(weekSymptoms.length, baselinePerWeek),
      distinct: byName.size,
      distribution: cards.map((c) => ({ name: c.name, catalogKey: c.catalogKey, count: c.occurrences, share: c.shareOfWeek })),
      cards,
      onsetWindows,
    },
  };
}

function daysBetween(a: string, b: string): number {
  return Math.round((Date.parse(b) - Date.parse(a)) / 86_400_000);
}

File: api/app/domain/digest.py
"""The weekly digest: how the week went, and per-symptom detail."""

import math
from dataclasses import dataclass, field
from typing import Literal

from app.domain.catalog import catalog_by_key, catalog_by_name
from app.domain.coverage import COVERAGE_SLOTS
from app.domain.flares import cluster_flares, intensity_of, onset_hours
from app.domain.rounding import js_round, round1, round2
from app.domain.rows import CorrelationRow, MealRow, SettingsRow, SymptomRow
from app.domain.severity import discomfort_score
from app.domain.time import add_days, days_between, local_date, week_dates, weekday_label

Level = Literal["zero", "moderate", "high"]
Trend = Literal["up", "same", "down"]

# Walking the baseline day by day is bounded so a stray old entry can't hang a request
MAX_BASELINE_DAYS = 3660


@dataclass
class DayTrend:
    date: str
    weekday: str
    index: float
    occurrences: int
    max_intensity: int
    level: Level


@dataclass
class SymptomCard:
    name: str
    catalog_key: str | None
    emoji: str | None
    occurrences: int
    share_of_week: float
    avg_severity10: float
    avg_duration_minutes: int | None
    peak_day: str | None
    vs_baseline: Trend
    top_triggers: list[str]


@dataclass
class OnsetWindows:
    under1h: int = 0
    from1to3h: int = 0
    over3h: int = 0
    unmatched: int = 0


@dataclass
class Trends:
    index: float
    baseline_index: float
    delta_vs_baseline_percent: int
    previous_week_index: float
    change_vs_previous_percent: int
    occurrences: int
    flares: int
    discomfort_free_days: int
    severe_days: int
    severe_peak_day: str | None
    meal_log_depth: float
    data_completeness: Literal["complete", "partial", "empty"]
    days: list[DayTrend]


@dataclass
class SymptomsSection:
    total: int
    baseline_per_week: float
    vs_baseline: Trend
    distinct: int
    distribution: list[dict]
    cards: list[SymptomCard]
    onset_windows: OnsetWindows


@dataclass
class WeeklyDigest:
    week_start: str
    week_end: str
    previous_week_start: str
    has_next_week: bool
    trends: Trends
    symptoms: SymptomsSection


def day_index(symptoms: list[SymptomRow]) -> float:
    """A day's discomfort on the 0-10 scale: its worst symptom, doubled."""
    return 0 if not symptoms else discomfort_score(max(intensity_of(s) for s in symptoms))


def level_for(index: float) -> Level:
    if index == 0:
        return "zero"
    return "high" if index >= 5 else "moderate"


def compare(value: float, baseline: float) -> Trend:
    if baseline == 0:
        return "same" if value == 0 else "up"
    ratio = value / baseline
    if ratio > 1.2:
        return "up"
    if ratio < 0.8:
        return "down"
    return "same"


def _percent_change(value: float, reference: float) -> int:
    return js_round((value - reference) / reference * 100) if reference > 0 else 0


def compute_weekly_digest(
    meals: list[MealRow],
    symptoms: list[SymptomRow],
    correlations: list[CorrelationRow],
    week_start: str,
    tz: str,
    settings: SettingsRow,
    today: str,
) -> WeeklyDigest:
    """`meals` and `symptoms` are the user's whole history: the week is scored
    against everything logged so far."""
    dates = week_dates(week_start)
    week_end = dates[6]

    symptoms_by_day: dict[str, list[SymptomRow]] = {}
    for symptom in symptoms:
        symptoms_by_day.setdefault(local_date(symptom.timestamp, tz), []).append(symptom)
    meals_by_day: dict[str, list[MealRow]] = {}
    for meal in meals:
        meals_by_day.setdefault(local_date(meal.timestamp, tz), []).append(meal)

    days = [
        DayTrend(
            date=date,
            weekday=weekday_label(date),
            index=day_index(symptoms_by_day.get(date, [])),
            occurrences=len(symptoms_by_day.get(date, [])),
            max_intensity=max(
                (intensity_of(s) for s in symptoms_by_day.get(date, [])), default=0
            ),
            level=level_for(day_index(symptoms_by_day.get(date, []))),
        )
        for date in dates
    ]
    # Always over seven days, so a partly-logged week reads as calmer, not worse
    week_index = sum(day.index for day in days) / 7

    logged_dates = sorted(set(symptoms_by_day) | set(meals_by_day))
    baseline_index = 0.0
    if logged_dates:
        cursor, last = logged_dates[0], max(today, week_end)
        total = count = 0
        while cursor <= last and count < MAX_BASELINE_DAYS:
            total += day_index(symptoms_by_day.get(cursor, []))
            count += 1
            cursor = add_days(cursor, 1)
        baseline_index = total / count if count else 0.0

    previous_week_start = add_days(week_start, -7)
    previous_index = sum(
        day_index(symptoms_by_day.get(date, [])) for date in week_dates(previous_week_start)
    ) / 7

    week_symptoms = [s for date in dates for s in symptoms_by_day.get(date, [])]
    peak = max(days, key=lambda d: d.index) if days else None

    slots_logged = sum(
        len({m.meal_type for m in meals_by_day.get(date, [])} & set(COVERAGE_SLOTS))
        for date in dates
    )
    logged_days = sum(
        1 for date in dates if meals_by_day.get(date) or symptoms_by_day.get(date)
    )
    elapsed_days = sum(1 for date in dates if date <= today)
    completeness = (
        "empty" if logged_days == 0 else "complete" if logged_days >= elapsed_days else "partial"
    )

    by_name: dict[str, list[SymptomRow]] = {}
    for symptom in week_symptoms:
        by_name.setdefault(symptom.name, []).append(symptom)

    history_weeks = max(
        1,
        math.ceil((days_between(logged_dates[0], today) + 1) / 7) if logged_dates else 1,
    )
    history_counts: dict[str, int] = {}
    for symptom in symptoms:
        history_counts[symptom.name] = history_counts.get(symptom.name, 0) + 1

    cards: list[SymptomCard] = []
    for name, entries in by_name.items():
        catalog = catalog_by_key(entries[0].catalog_key) or catalog_by_name(name)
        durations = [s.duration_minutes for s in entries if s.duration_minutes is not None]
        per_day: dict[str, int] = {}
        for symptom in entries:
            date = local_date(symptom.timestamp, tz)
            per_day[date] = per_day.get(date, 0) + 1
        peak_day = min(
            sorted(per_day), key=lambda date: -per_day[date], default=None
        ) if per_day else None
        cards.append(
            SymptomCard(
                name=name,
                catalog_key=entries[0].catalog_key or (catalog.key if catalog else None),
                emoji=catalog.emoji if catalog else None,
                occurrences=len(entries),
                share_of_week=round1(len(entries) / len(week_symptoms) * 100) / 100,
                avg_severity10=round1(
                    sum(discomfort_score(intensity_of(s)) for s in entries) / len(entries)
                ),
                avg_duration_minutes=js_round(sum(durations) / len(durations)) if durations else None,
                peak_day=peak_day,
                vs_baseline=compare(len(entries), history_counts.get(name, 0) / history_weeks),
                top_triggers=[
                    row.food_name
                    for row in sorted(
                        (
                            r for r in correlations
                            if r.symptom_name == name and r.dimension == "ingredient"
                        ),
                        key=lambda r: -r.confidence,
                    )
                ][:3],
            )
        )
    cards.sort(key=lambda c: (-c.occurrences, c.name.casefold(), c.name))

    windows = OnsetWindows()
    for symptom in week_symptoms:
        hours = onset_hours(meals, symptom.timestamp, settings.correlation_window_hours)
        if hours is None:
            windows.unmatched += 1
        elif hours < 1:
            windows.under1h += 1
        elif hours <= 3:
            windows.from1to3h += 1
        else:
            windows.over3h += 1

    baseline_per_week = len(symptoms) / history_weeks
    return WeeklyDigest(
        week_start=week_start,
        week_end=week_end,
        previous_week_start=previous_week_start,
        has_next_week=add_days(week_start, 7) <= today,
        trends=Trends(
            index=round1(week_index),
            baseline_index=round1(baseline_index),
            delta_vs_baseline_percent=_percent_change(week_index, baseline_index),
            previous_week_index=round1(previous_index),
            change_vs_previous_percent=_percent_change(week_index, previous_index),
            occurrences=len(week_symptoms),
            flares=len(cluster_flares(week_symptoms)),
            discomfort_free_days=sum(1 for day in days if day.index == 0),
            severe_days=sum(1 for day in days if day.level == "high"),
            severe_peak_day=peak.date if peak and peak.index > 0 else None,
            meal_log_depth=round2(slots_logged / 21),
            data_completeness=completeness,
            days=days,
        ),
        symptoms=SymptomsSection(
            total=len(week_symptoms),
            baseline_per_week=round1(baseline_per_week),
            vs_baseline=compare(len(week_symptoms), baseline_per_week),
            distinct=len(by_name),
            distribution=[
                {
                    # Hand-built payload, so the client's spelling is used directly
                    "name": card.name,
                    "catalogKey": card.catalog_key,
                    "count": card.occurrences,
                    "share": card.share_of_week,
                }
                for card in cards
            ],
            cards=cards,
            onset_windows=windows,
        ),
    )
